air/src/execution_step/boxed_value/stream_map.rs

```diff
@@ -81,32 +81,18 @@ impl StreamMap {
         &mut self.stream
     }
 
-    // TODO: change the implementation to mutate the underlying stream
-    // instead of creating a new one
-    pub(crate) fn create_unique_keys_stream(&self) -> Stream {
+    /// Returns an iterator to values with unique keys.
+    pub(crate) fn iter_unique_key(&self) -> impl Iterator<Item = &ValueAggregate> {
         use std::collections::HashSet;
 
         let mut met_keys = HashSet::new();
 
-        // unwrap is safe because slice_iter always returns Some if supplied generations are valid
-        let new_values = self
-            .stream
-            .slice_iter(Generation::Nth(0.into()), Generation::Last)
-            .unwrap()
-            .map(|values| {
-                values
-                    .iter()
-                    .filter(|v| {
-                        StreamMapKey::from_kvpair(v)
-                            .map(|key| met_keys.insert(key))
-                            .unwrap_or(false)
-                    })
-                    .cloned()
-                    .collect::<Vec<_>>()
-            })
-            .collect::<Vec<_>>();
-
-        Stream::new(new_values, self.stream.previous_gens_count())
+        // it's always possible to go through all values
+        self.stream.iter(Generation::Last).unwrap().filter(move |value| {
+            StreamMapKey::from_kvpair(value)
```

Comment:
`...::from_kvpair` can be optimized, returning `Cow::Borrowed`.

Reply:
good catch, but lets do it in a separate PR

Reply:
No problem.